### PR TITLE
[Feature] Grading System - Absolute Points for Grade Key

### PR DESCRIPTION
**Pull request by ivohashamov**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Previously, instructors could only create a grade key by setting the boundaries of the grade steps in percentages. For convenience now, they can also set the grade key using absolute points, as long as the max points property for the course/exam has been set.

### Description

I've created a max points field for the course which servers a similar purpose as the one for the exam and is necessary for the other changes in this PR.
I've added a new input field on the grade key page, which sets and displays the max points for courses/exams. Furthermore, the grade step table has 2 additional columns for the minimum and maximum points. Changing either of the values of a boundary (the percentage or the points) automatically adapts the other value. Also, changing the max points of the course/exam maintains the percentage values, but adapts the absolute points values accordingly. 
It's worth mentioning that the grade key matching still relies solely on the percentage values, and the possibility to add the steps with absolute points is only there for convenience.

### Steps for Testing

1. Log in to Artemis as an Instructor (Using TS3 and my course "IvoTC" could save you some time, I have an exam there with the necessary student submissions for testing)
2. Navigate to Course Administration -> Grade Key
3. Try entering different combinations of values.
- if max points for the course aren't set, the min/max points fields should be empty and disabled, but valid grade keys should still be saveable
- updating the points/percentages for a lower/upper bound should update the percentages/points for that same bound, such that (percentage * max points) / 100 = points always holds (with an epsilon margin of error)
- updating the max points, should automatically adapt all grade steps, where the percentages are kept, and the points are adjusted accordingly
4. Go to the course edit page -> there should be a warning next to the max points fields regarding changing the value
5. Try updating the value and saving it -> then go back to the grade key page -> the value should be adjusted to the one in the edit page and all of the grade steps should have been adjusted as in step 3 (percentages maintained/points adapted)
6. Repeat the steps accordingly for an exam and its grade key
7. Finally, find an exam with student submissions and published results and create a grade key for it. Then log in as one of the students with submission in this exam and go to the exam overview page -> Click on View Grade Key -> the intervals should be displayed in points as well.

### Test Coverage

GradingScaleService.java: 94%
GradingScaleRepository.java: 100%
grade-key-overview.component.ts: 97.22%
grading-system.component.ts: 94.25%

### Screenshots

Grade key page:
Screenshot 2021-06-14 162937
Course/Exam edit page:
Screenshot 2021-06-14 163018
Grade Key Overview for exams:
image

### CODE REVIEW

**Review by canyelten — CHANGES_REQUESTED** [review]
Tested on TS2, found the following naming problem:

**Comment by ivohashamov** [comment]
> Tested on TS2, found the following naming problem:
> 
> 
> 

That's odd, I just tested the latest build on TS2 and didn't get that issue. Locally it's also running fine. Did you deploy the latest build?

**Review by ArtemTsurikov — CHANGES_REQUESTED** [review]
I tested it on TS1. The functionalities work well, but there are some issues with diskplaying the titles and labels. Furthermore there is the problem that if you include half points, the field is too small and not everything is displayed (see screenshot 2).
Screenshot 2021-06-17 at 11 31 39
Screenshot 2021-06-17 at 11 29 15

**Comment by ivohashamov** [comment]
I'm officially confused. I can't reproduce the issues mentioned on any of the test servers. All seems to be working fine both on them and locally. I also double-checked the translation paths and they are fine. Can you try refreshing the page when testing, maybe it takes some time to load?

**Review by ArtemTsurikov — APPROVED** [review]
Found the problem. The titles and labels were not displayed, due to old caches. Deleting the cashes solves the issue. Approve!

File: src/main/webapp/i18n/de/exam.json [hunk 1]
```diff
@@ -241,7 +242,10 @@
             "examStudentReviewStart": "Beginn der Klausureinsicht",
             "examStudentReviewEnd": "Ende der Klausureinsicht",
             "studentReviewEnabled": "Klausureinsicht ist offen",
-            "maxPoints": "Anzahl an erreichbaren Punkten",
+            "maxPoints": {
+                "title": "Anzahl an erreichbaren Punkten",
+                "warning": "Durch die Aktualisierung dieses Feldes wird der Notenschlüssel für die Prüfung aktualisiert, so dass die Punktwerte für die Notenstufen weiterhin den Prozentsätzen entsprechen, angepasst an den neuen maximalen Punktwert"
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/exam.json:247]: sodass

File: src/main/webapp/i18n/de/gradingSystem.json [hunk 2]
```diff
@@ -30,6 +32,10 @@
                     "title": "Allgemeine Informationen",
                     "explanation": "Ein Notenschlüssel besteht aus Notenstufen, die alle möglichen Prozentwerte von 0 bis 100 abdecken. Alle benachbarten Notenstufen sollten sich an ihren Schranken überschneiden (Beispiel: Notenstufen (60, 70) und (70, 80) überschneiden sich bei 70%). Welche Notenstufe diese Schranke einschließt, wird durch die \"Inklusivität\" bestimmt. Alle Notenstufen haben außerdem entweder einen Namen (1,0, Ausreichend, A,...) oder eine Bonuspunktzahl (5, 10,..), abhängig von dem \"Notentyp\""
                 },
+                "maxPoints": {
+                    "title": "Maximale Punkte",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/gradingSystem.json:36]: Maximale Punktzahl

File: src/main/webapp/i18n/de/course.json [hunk 3]
```diff
@@ -61,6 +61,10 @@
             "startDate": "Start",
             "endDate": "Ende",
             "semester": "Semester",
+            "maxPoints": {
+                "title": "Maximale Punkte für den Kurs",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/course.json]: Maximale Punktzahl

File: src/main/webapp/i18n/de/course.json [hunk 4]
```diff
@@ -61,6 +61,10 @@
...
+                "warning": "Durch die Aktualisierung dieses Feldes wird der Notenschlüssel für den Kurs aktualisiert, so dass die Punktwerte für die Notenstufen weiterhin den Prozentsätzen entsprechen, angepasst an den neuen maximalen Punktwert"
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/course.json:66]: sodass

File: src/main/webapp/i18n/de/exam.json [hunk 5]
```diff
@@ -241,7 +242,10 @@
             "examStudentReviewStart": "Beginn der Klausureinsicht",
             "examStudentReviewEnd": "Ende der Klausureinsicht",
             "studentReviewEnabled": "Klausureinsicht ist offen",
-            "maxPoints": "Anzahl an erreichbaren Punkten",
+            "maxPoints": {
+                "title": "Anzahl an erreichbaren Punkten",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/exam.json:246]: Erreichbare Punktzahl

File: src/main/webapp/i18n/de/gradingSystem.json [hunk 6]
```diff
@@ -17,7 +17,9 @@
             },
             "firstPassingGrade": "Erste bestehende Note",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/de/gradingSystem.json]: Bestanden ab?

**Review by lschlesinger — CHANGES_REQUESTED** [review]
I found some more minor translations improvements. Also I would not use the word "Punktwert" in German but rather "Punktzahl" - if you agree it has to be changed at various places.

**Review by lschlesinger — APPROVED** [review]
Tested on TS3 according to your instructions. Functionality works, some nice-to-have features (such as connecting max and min values for adjacent grade steps) might be added later on.

**Review by SimonEntholzer — APPROVED** [review]
Worked while testing on TS3 during testing session

**Review by JohannesStoehr — DISMISSED** [review]
Worked while testing on TS3 during testing session

**Review by tobiasjungmann — DISMISSED** [review]
Tested in the testing session on ts3, worked as expected.

**Review by imstreetspirit — DISMISSED** [review]
tested on ts3, code also seems good

**Review by ndricimrr — DISMISSED** [review]
Tested on TS3. Working well

**Review by MDK5 — DISMISSED** [review]
Tested in testing session on ts3, worked well, but it would be a great addition for a follow-up to add a button to calculate the course/exam max points as sum of the exercise points and/or ensure the max points cannot be lower than that sum.

**Review by maximilianbergmann — DISMISSED** [review]
Tested on TS3. Works as expected

**Review by canyelten — APPROVED** [review]
Works as expected, deleting the cashes solved the issue.

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 7]
```diff
@@ -561,4 +564,12 @@ public String getCourseArchivePath() {
     public void setCourseArchivePath(String courseArchiveUrl) {
         this.courseArchivePath = courseArchiveUrl;
     }
+
+    public int getMaxPoints() {
+        return maxPoints != null ? maxPoints : 0;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Course.java:569]: I'm not sure if this implementation is good. In my opinion there could be a difference between max points being null (= not set) and max points being 0

File: src/main/java/de/tum/in/www1/artemis/domain/GradeStep.java [hunk 8]
```diff
@@ -25,12 +25,18 @@
     @Column(name = "lower_bound_percentage")
     private double lowerBoundPercentage;
 
+    @Column(name = "lower_bound_points")
+    private Double lowerBoundPoints;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/GradeStep.java:29]: Why do we store those values now in the database? I thought the calculation is only based on the percentage values and absolute values will only be used in the user interface to prevent redundant information in the database?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 9]
```diff
@@ -214,6 +219,19 @@ public ExamResource(UserRepository userRepository, CourseRepository courseReposi
         return ResponseEntity.ok().headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, result.getTitle())).body(result);
     }
 
+    private void handleMaxPointUpdates(Long examId, int existingMaxPoints, int updatedMaxPoints) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:222]: This method mostly includes duplicated code. Please avoid this

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 10]
```diff
@@ -214,6 +219,19 @@ public ExamResource(UserRepository userRepository, CourseRepository courseReposi
...
+        if (updatedMaxPoints < 0) {
+            throw new BadRequestAlertException("Max points must be greater than 0", ENTITY_NAME, "invalidMaxPoints", true);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:224]: The message is not accurate. 
What about the case `updatedMaxPoints` = 0?

File: src/main/webapp/app/course/manage/course-update.component.ts [hunk 11]
```diff
@@ -124,6 +141,9 @@ export class CourseUpdateComponent implements OnInit {
                 onlineCourse: new FormControl(this.course.onlineCourse),
                 complaintsEnabled: new FormControl(this.complaintsEnabled),
                 requestMoreFeedbackEnabled: new FormControl(this.requestMoreFeedbackEnabled),
+                maxPoints: new FormControl(this.course.maxPoints, {
+                    validators: [Validators.min(1)],
```
Comment (krusche) [inline src/main/webapp/app/course/manage/course-update.component.ts:145]: is this validator only active when the value is set?

File: src/main/webapp/app/course/manage/course-update.component.html [hunk 12]
```diff
@@ -10,7 +10,7 @@ <h4 *ngIf="course.id" id="jhi-course-heading-edit" jhiTranslate="artemisApp.cour
             <hr />
             <div>
                 <jhi-alert></jhi-alert>
-                <jhi-alert-error></jhi-alert-error>
+                <jhi-alert-error *ngIf="!gradingScaleNotFoundForCourse"></jhi-alert-error>
```
Comment (krusche) [inline src/main/webapp/app/course/manage/course-update.component.html:13]: I don't understand this change.

File: src/main/webapp/app/course/manage/course-update.component.ts [hunk 13]
```diff
@@ -47,6 +48,8 @@ export class CourseUpdateComponent implements OnInit {
     customizeGroupNames = false; // default value
     presentationScorePattern = /^[0-9]{0,4}$/; // makes sure that the presentation score is a positive natural integer greater than 0 and not too large
     courseOrganizations: Organization[];
+    gradingScaleExistsForCourse = false;
```
Comment (krusche) [inline src/main/webapp/app/course/manage/course-update.component.ts:51]: why do we need those two variables?

File: src/main/webapp/app/exam/manage/exams/exam-update.component.ts [hunk 14]
```diff
@@ -18,13 +19,16 @@ export class ExamUpdateComponent implements OnInit {
     exam: Exam;
     course: Course;
     isSaving: boolean;
+    gradingScaleExistsForExam = false;
```
Comment (krusche) [inline src/main/webapp/app/exam/manage/exams/exam-update.component.ts:22]: same as above

File: src/main/webapp/app/exam/manage/exams/exam-update.component.html [hunk 15]
```diff
@@ -10,7 +10,7 @@ <h4>{{ 'artemisApp.examManagement.editExam' | artemisTranslate }}</h4>
             </ng-template>
 
             <div>
-                <jhi-alert-error></jhi-alert-error>
+                <jhi-alert-error *ngIf="!gradingScaleNotFoundForExam"></jhi-alert-error>
```
Comment (krusche) [inline src/main/webapp/app/exam/manage/exams/exam-update.component.html:13]: same as above

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 16]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
...
+                        <input [(ngModel)]="gradeStep.upperBoundPercentage" type="number" step="0.01" min="0" max="100" required (change)="setPoints(gradeStep, false)" />
                     </td>
                     <td>
-                        <input [(ngModel)]="gradeStep.upperBoundPercentage" type="number" min="0" max="100" required />
+                        <input
+                            [(ngModel)]="gradeStep.upperBoundPoints"
+                            type="number"
+                            step="0.01"
+                            min="0"
+                            [max]="getMaxPoints() || 999"
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:107]: 999 looks like a magic number

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 17]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
...
                     <td>
-                        <input [(ngModel)]="gradeStep.lowerBoundPercentage" type="number" min="0" max="100" required />
+                        <input [(ngModel)]="gradeStep.upperBoundPercentage" type="number" step="0.01" min="0" max="100" required (change)="setPoints(gradeStep, false)" />
                     </td>
                     <td>
-                        <input [(ngModel)]="gradeStep.upperBoundPercentage" type="number" min="0" max="100" required />
+                        <input
+                            [(ngModel)]="gradeStep.upperBoundPoints"
+                            type="number"
+                            step="0.01"
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:105]: 0.01 should only be used for rounding. I don't think using it as step value makes a lot of sense

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 18]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
...
+                        <input [(ngModel)]="gradeStep.lowerBoundPercentage" type="number" step="0.01" min="0" max="100" required (change)="setPoints(gradeStep, true)" />
+                    </td>
+                    <td>
+                        <input
+                            [(ngModel)]="gradeStep.lowerBoundPoints"
+                            type="number"
+                            step="0.01"
+                            min="0"
+                            [max]="getMaxPoints() || 999"
+                            [required]="maxPointsValid()"
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:93]: you have a lot of method calls here which can make the whole page very slow due to Angular's automatic change tracking. I suggest to reduce the number of method calls or to change the change tracking strategy (see e.g. the quiz edit page)

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 19]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
                 </tr>
             </thead>
             <tbody>
-                <tr *ngFor="let gradeStep of gradingScale.gradeSteps; let i = index">
+                <tr *ngFor="let gradeStep of gradingScale.gradeSteps; let i = index" class="grading-scale-table-row">
+                    <td>
+                        <input [(ngModel)]="gradeStep.lowerBoundPercentage" type="number" step="0.01" min="0" max="100" required (change)="setPoints(gradeStep, true)" />
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:84]: I would use a step of 0.5 here

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 20]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
...
+                            min="0"
+                            [max]="getMaxPoints() || 999"
+                            [required]="maxPointsValid()"
+                            [disabled]="!maxPointsValid()"
+                            (change)="setPercentage(gradeStep, true)"
+                        />
+                    </td>
                     <td>
-                        <input [(ngModel)]="gradeStep.lowerBoundPercentage" type="number" min="0" max="100" required />
+                        <input [(ngModel)]="gradeStep.upperBoundPercentage" type="number" step="0.01" min="0" max="100" required (change)="setPoints(gradeStep, false)" />
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:99]: I would use a step of 0.5 here

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 21]
```diff
@@ -57,21 +79,45 @@ <h2 style="margin-bottom: 0.5em">
...
+                            [disabled]="!maxPointsValid()"
+                            (change)="setPercentage(gradeStep, false)"
+                        />
                     </td>
                     <td *ngIf="isGradeType()">
                         <input [(ngModel)]="gradeStep.gradeName" type="text" required />
                     </td>
                     <td *ngIf="!isGradeType()">
-                        <input [(ngModel)]="gradeStep.gradeName" type="number" step="0.01" min="0" placeholder="0" required />
+                        <input [(ngModel)]="gradeStep.gradeName" type="number" step="0.25" min="0" placeholder="0" required />
```
Comment (krusche) [inline src/main/webapp/app/grading-system/grading-system.component.html:117]: I don't understand this. Why do we use number in case it is not a grade type? (the method name is not very intuitive, what is a grade type?)

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java (hunk 7 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java:569]: @Nullable

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 22]
```diff
@@ -214,6 +219,19 @@ public ExamResource(UserRepository userRepository, CourseRepository courseReposi
...
+        }
+        var gradingScale = this.gradingScaleRepository.findByExamId(examId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:226]: findByExamIdElseThrow(...)

Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:226]: please use a proper type

**Review by KiraNickel — DISMISSED** [review]
Tested on ts3, works as described 👍🏼

**Review by cremertim — APPROVED** [review]
Works on ts3

File: src/main/webapp/app/course/manage/course-update.component.ts (hunk 11 above)
Comment (ivohashamov) [reply src/main/webapp/app/course/manage/course-update.component.ts:128]: yes

File: src/main/webapp/app/grading-system/grading-system.component.html (hunk 21 above)
Comment (ivohashamov) [reply src/main/webapp/app/grading-system/grading-system.component.html:115]: I removed the method, now it just checks if  `gradingScale.gradeType != GradeType.GRADE`. Because the bonus points are a number, and I wanted to use the built-in validation to enforce that.

**Review by SimonEntholzer — APPROVED** [review]
Tested on TS1 :+1:  works as expected

**Review by imstreetspirit — APPROVED** [review]
tested on ts1 again. works as described

**Review by ndricimrr — APPROVED** [review]
Tested again on TS1. Working as expected.

File: src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java [hunk 23]
```diff
@@ -84,22 +84,26 @@ public GradeStepResource(GradingScaleRepository gradingScaleRepository, GradeSte
...
+        if (!isInstructor && !exam.resultsPublished()) {
             return forbidden();
         }
-        GradeStepsDTO gradeStepsDTO = prepareGradeStepsDTO(gradingScale);
+        GradeStepsDTO gradeStepsDTO = prepareGradeStepsDTO(gradingScale, exam.getMaxPoints());
         return ResponseEntity.ok(gradeStepsDTO);
     }
 
-    private GradeStepsDTO prepareGradeStepsDTO(GradingScale gradingScale) {
+    private GradeStepsDTO prepareGradeStepsDTO(GradingScale gradingScale, int maxPoints) {
```
Comment (lschlesinger) [inline src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java:98]: question: isn't it `Integer` as declared in `src/main/java/de/tum/in/www1/artemis/domain/Course.java` and in `src/main/java/de/tum/in/www1/artemis/web/rest/dto/GradeStepsDTO.java`

Comment (ivohashamov) [reply src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java:98]: because the getMaxPoints in the Exam.java returns an int (it's apparently no longer optional). However, it's optional in the Course, hence the Integer there. I kept the Integer in the GradeStepsDTO since I will reuse it, in my next PR, where the equivalent Course endpoint would return the GradeStepsDTO. Since it can be optional for the course I opted to have it as optional in the DTO in general and just always set it for the Exam.

File: src/main/webapp/i18n/en/gradingSystem.json [hunk 24]
```diff
@@ -45,8 +51,10 @@
             },
             "error": {
                 "empty": "Grade steps are empty!",
+                "negativeMaxPoints": "Maximum points can't be negative",
                 "emptyFields": "All grade step fields must be filled out!",
```
Comment (lschlesinger) [inline src/main/webapp/i18n/en/gradingSystem.json:55]: Not in the scope of this PR but still relevant for the next time you work on the grade steps: this should be changed to either "...must be filled" or "...must be filled in" (AFAIK filled out is not correct)

File: src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java (hunk 23 above)
Comment (lschlesinger) [reply src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java:98]: ah ok, it is obviously not related to the functionality provided in that PR but I did not 100% get your explanation... why would it depend on a value being optional? One could also change the return type of the `getMaxPoints()` in `Exam.java` to Integer.

**Comment by lschlesinger** [comment]
Is there a reason why min/max % is still not restricted on 2 decimals while min/max points is? The provided precision is not needed I guess...

Bildschirmfoto 2021-06-18 um 22 26 49

**Review by lschlesinger — COMMENTED** [review]
Is there a reason why min/max % is still not restricted on 2 decimals while min/max points is? The provided precision is not needed I guess...

Bildschirmfoto 2021-06-18 um 22 26 49

File: src/main/resources/config/liquibase/changelog/20210618160000_changelog.xml [hunk 25]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.9.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+    <changeSet author="hashamov" id="20210618160000">
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20210618160000_changelog.xml:3]: Please remove this file. It's not a good idea to first add columns and then delete them. This would screw up production server with unnecessary change sets

File: src/main/resources/config/liquibase/changelog/20210605180001_changelog.xml [hunk 26]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.9.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.9.xsd">
+    <changeSet author="hashamov" id="20210605180001">
+        <addColumn tableName="grade_step">
+            <column name="lower_bound_points" type="DOUBLE"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20210605180001_changelog.xml:5]: please delete those two lines, so that in total only `course.max_points` will be added

File: src/main/resources/config/liquibase/master.xml [hunk 27]
```diff
@@ -122,6 +122,8 @@
     <include file="classpath:config/liquibase/changelog/20210519162242_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20210528104100_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20210528163200_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20210605180001_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20210618160000_changelog.xml" relativeToChangelogFile="false"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/master.xml:126]: delete

**Review by krusche — CHANGES_REQUESTED** [review]
please make sure that we only have one change set and not two!

File: src/main/resources/config/liquibase/changelog/20210605180001_changelog.xml (hunk 26 above)
Comment (ivohashamov) [reply src/main/resources/config/liquibase/changelog/20210605180001_changelog.xml:5]: I would have. However, since the changes were already executed on multiple test servers, I created another one, so there are no obsolete columns. Just deleting the lines would work fine on production, but it would screw up the test servers because of the checksum.

File: src/main/resources/config/liquibase/master.xml (hunk 27 above)
Comment (ivohashamov) [reply src/main/resources/config/liquibase/master.xml:126]: > Is there a reason why min/max % is still not restricted to 2 decimals while min/max points is? The provided precision is not needed I guess...
> 
> Bildschirmfoto 2021-06-18 um 22 26 49

neither is restricted to 2 decimal points. I'm not rounding the values on this page, so the instructor is exactly aware of what the grades will be compared against. We wanted to keep the precision

**Comment by ivohashamov** [comment]
> Is there a reason why min/max % is still not restricted to 2 decimals while min/max points is? The provided precision is not needed I guess...
> 
> Bildschirmfoto 2021-06-18 um 22 26 49

Neither are restricted to 2 decimal places. There is no rounding on the page, so the instructor is exactly aware of what the grades would be compared against. We wanted to keep the precision.

File: src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java (hunk 23 above)
Comment (ivohashamov) [reply src/main/java/de/tum/in/www1/artemis/web/rest/GradeStepResource.java:98]: One could do that, true. However, since it's not an optional value anymore, it doesn't make sense for it to return an Integer. Therefore, I left it as int.

File: src/main/webapp/i18n/en/gradingSystem.json (hunk 24 above)
Comment (ivohashamov) [reply src/main/webapp/i18n/en/gradingSystem.json:55]: Could be. I know there's the "fill out a form" phrase in English.  Haven't really looked that up explicitly when I wrote this.

File: src/main/resources/config/liquibase/changelog/20210618160000_changelog.xml (hunk 25 above)
Comment (ivohashamov) [reply src/main/resources/config/liquibase/changelog/20210618160000_changelog.xml:3]: I agree. I'm not sure how I can revert the changes on the test servers otherwise though.

File: src/main/resources/config/liquibase/master.xml (hunk 27 above)
Comment (ivohashamov) [reply src/main/resources/config/liquibase/master.xml:126]: If I do that, there would be obsolete columns on the test servers and production

File: src/main/webapp/app/exam/participate/summary/points-summary/grade-key-overview/grade-key-overview.component.ts [hunk 28]
```diff
@@ -43,11 +46,41 @@ export class GradeKeyOverviewComponent implements OnInit {
...
+    /**
+     * Rounds a number to two decimal places
+     *
+     * @param num the number to be rounded
+     */
+    round(num?: number) {
+        if (num == undefined) {
+            return;
+        }
+        return Math.round(num * 100) / 100;
```
Comment (krusche) [inline src/main/webapp/app/exam/participate/summary/points-summary/grade-key-overview/grade-key-overview.component.ts:84]: please use the existing rounding methods in `utils.ts` and do not introduce new ones.
